### DISCUSSION TITLE
Expose the response body text for plugins

### DISCUSSION THREAD

**Discussion by iandunn** [Kong/insomnia#3571] [Ideas]
### Problem

#423 requested being able to get and set the request and response bodies. #2003 added that ability for requests, but not  for responses. #1830 was related.

My use case is that I'd like to write a plugin that appends a formatted date/time to timestamp values in  responses, so that they're human-readable. For example, `1599858469` would become `1599858469 (Fri Sep 11 14:07:49 PDT 2020)`.

It seems like the only current method is working with the buffers, which requires jumping through extra hoops.

### Ideal solution

Add `getBodyText()` and `setBodyText()` functions.

### Alternate solution

Add examples of working with buffers to the documentation, so developers can just copy/paste the code.

**Comment by jeremiahsmall** [comment]
We have a very common use case where the response is JSON containing a single property we're primarily interested in which is escaped JSON. We love insomnia-plugin-inception for requests, and it would be awesome if we could also apply similar to responses. Perhaps this would also mean supporting plugins via jsonpath syntax?

**Reply by develohpanda** [reply]
Wow I just learned about [insomnia-plugin-inception](https://insomnia.rest/plugins/insomnia-plugin-inception). Plugins continue to blow my mind! This is super cool!

It's unlikely that we'll add to JSONPath to support plugins, but there are a few ideas I have in mind to achieve this via a plugin. Would the suggested solution of documentation help with your use case as well?

**Comment by develohpanda** [comment]
Hi @iandunn! I just saw this now, apologies for the delays. In case this is still useful, here's an example (which can move into docs once we find a good place for it).

```js
// For help writing plugins, visit the documentation to get started:
//   https://support.insomnia.rest/article/173-plugins

const bufferToJsonObj = buf => JSON.parse(buf.toString('utf-8'));
const jsonObjToBuffer = obj => Buffer.from(JSON.stringify(obj), 'utf-8');

module.exports.responseHooks = [
  async ctx => {
    try{
      const resp = bufferToJsonObj(ctx.response.getBody());
      
      // Modify
      resp.__randomNumber = Math.random();

      // If you want something from a user, use prompt for example
      const promptResult = await ctx.app.prompt('Type somthing', { defaultValue: 'abcd' });
      resp.__customValue = promptResult;

      ctx.response.setBody(jsonObjToBuffer(resp));
    } catch {
      // no-op
    }
  }
]
```

Here's a loom recording showing this working: https://www.loom.com/share/35e63deec05344c595d95544857b70a5.

**Reply by iandunn** [reply]
Thanks :)

**Reply by falondarville** [reply]
We've updated the docs! https://docs.insomnia.rest/insomnia/introduction-to-plugins#alter-a-response-body

**Reply by jeremiahsmall** [reply]
This example worked perfectly for our needs. We published [insomnia-plugin-filemaker](https://insomnia.rest/plugins/insomnia-plugin-filemaker) which automatically unescapes the script result property for us. Big time saver. We didn't write a readme yet, because it's basically scratching our own itch for now, but eventually we'll add that and maybe a few other time-savers when dealing with FileMaker Data API.

**Reply by develohpanda** [reply]
Very cool! Thank you for sharing that 😄 Just going to make a timely plug here for the [Kong Summit Hackathon](https://github.com/Kong/insomnia/discussions/4016) that's on _next week_, and developers making Insomnia plugins can take part in it!
